### PR TITLE
Improve error handling for misnamed formula class

### PR DESCRIPTION
**Pull request by UniqMartin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [x] Have you successfully ran `brew tests` with your changes locally?

----

Improve on the confusing error behavior observed in #82, i.e. for the case where there is a mismatch between the expected and actual class name in a formula file and adapt `brew install` to suppress the confusing search in this case. (The new exception class is a subclass of `FormulaUnavailableError`, thus the change is transparent to other users of `Formulary` that haven't been adapted.)

Old behavior:

```
$ brew info qt5-webkit
Error: No available formula with the name "qt5-webkit"

$ brew install qt5-webkit
Error: No available formula with the name "qt5-webkit"
==> Searching for similarly named formulae...
This similarly named formula was found:
qt5-webkit
To install it, run:
  brew install qt5-webkit
==> Searching taps...
Error: No formulae found in taps.
```

New behavior:

```
$ brew info qt5-webkit
Error: No available formula with the name "qt5-webkit"
In formula file: /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/qt5-webkit.rb
Expected to find class Qt5Webkit, but only found: Qt5WebKit.

$ brew install qt5-webkit
Error: No available formula with the name "qt5-webkit"
In formula file: /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/qt5-webkit.rb
Expected to find class Qt5Webkit, but only found: Qt5WebKit.
```

New behavior when there's no class derived from `Formula`:

```
$ brew info qt5-webkit
Error: No available formula with the name "qt5-webkit"
In formula file: /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/qt5-webkit.rb
Expected to find class Qt5Webkit, but only found: SomeUnrelatedClass (not derived from Formula!).
```

New behavior when there are no classes in the formula file:

```
$ brew info qt5-webkit
Error: No available formula with the name "qt5-webkit"
In formula file: /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/qt5-webkit.rb
Expected to find class Qt5Webkit, but found no classes.
```

cc @xu-cheng

### CODE REVIEW

**Comment by xu-cheng** [comment]
LGTM, have you test it under Ruby 1.8?

**Comment by UniqMartin** [comment]
> have you test it under Ruby 1.8?

Yes, I have. (The new code is also covered by the tests and they succeeded on Maverikcks with 1.8.7.)

**Comment by The-Compiler** [comment]
Thanks for the quick fix! I can't test it right now, but the new behaviour definitely makes sense!

File: Library/Homebrew/cmd/install.rb [hunk 1]
```diff
@@ -139,6 +139,11 @@ def install
       perform_preinstall_checks
 
       formulae.each { |f| install_formula(f) }
+    rescue FormulaClassUnavailableError => e
+      # Need to re-raise before `FormulaUnavailableError` (superclass of this)
+      # is handled, as searching for a formula doesn't make sense here (the
+      # formula was found, but there's a problem with its implementation).
+      raise e
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/install.rb]: I think this should be `ofail e.meesage` as you should not raise and interrupt another formula's installation. Noted there is a loop here.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/install.rb]: 👍

**Comment by mikemcquaid** [comment]
Agreed with @xu-cheng's comment but otherwise 👍

File: Library/Homebrew/cmd/install.rb (hunk 1 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/install.rb]: This was a silly oversight. Thanks for noticing and pointing it out! :bow:

**Pull request closed without merging**